creme/stream/datasets/agrawal.py

```diff
@@ -0,0 +1,316 @@
+    {'salary': 98980.83074718699, 'commission': 0, 'age': 40, 'elevel': 0, 'car': 6, 'zipcode': 1, 'hvalue': 1159108.4298026664, 'hyears': 28, 'loan': 281644.10892276966} 1
+
+    References
+    ----------
+    [^1]: Rakesh Agrawal, Tomasz Imielinksi, and Arun Swami. "Database
+          Mining: A Performance Perspective", IEEE Transactions on Knowledge and
+          Data Engineering, 5(6), December 1993.
+
+    """
+    def __init__(self, classification_function=0,
+                 random_state=None,
+                 balance_classes=False,
+                 perturbation=0.0):
+        super().__init__(n_features=9, n_classes=2, n_outputs=1, task=base.BINARY_CLF)
+
+        # Classification functions to use
+        self._classification_functions = [self._classification_function_0,
+                                          self._classification_function_1,
+                                          self._classification_function_2,
+                                          self._classification_function_3,
+                                          self._classification_function_4,
+                                          self._classification_function_5,
+                                          self._classification_function_6,
+                                          self._classification_function_7,
+                                          self._classification_function_8,
+                                          self._classification_function_9]
+        if classification_function not in range(10):
+            raise ValueError(f"classification_function takes values from 0 to 9 "
+                             f"and {classification_function} was passed")
+        self.classification_function = classification_function
+        self.balance_classes = balance_classes
+        if not 0.0 <= perturbation <= 1.0:
+            raise ValueError(f"noise percentage should be in [0.0..1.0] "
+                             f"and {perturbation} was passed")
+        self.perturbation = perturbation
+        self.random_state = random_state
+        self._random_state = None  # This is the actual random_state object used internally
+        self.n_num_features = 6
+        self.n_cat_features = 3
+        self._next_class_should_be_zero = False
+        self.target_names = ["target"]
+        self.feature_names = ["salary", "commission", "age", "elevel", "car", "zipcode", "hvalue",
+                              "hyears", "loan"]
+        self.target_values = [i for i in range(self.n_classes)]
+        # Legacy variables
+        self.current_sample_x = None
+        self.current_sample_y = None
+        self.sample_idx = 0
+
+        self._prepare_for_use()
+
+    def _prepare_for_use(self):
+        self._random_state = check_random_state(self.random_state)
+        self._next_class_should_be_zero = False
+
+    def restart(self):
```

Comment:
I'm not sure about having a `restart` function. If we do have it, then do we need to implement for each generator?

I believe that an alternative approach would be to have to be `__iter__` be idempotent. That is, if you set a seed, then looping over the dataset a second time should produce the same data as when iterating over it a first time. In the case of this generator, this simply requires to set everything up at the start of `__iter__`.

What do you think? If we agree on this convention I'll have to modify the two existing generators, which are `SEA` and `Friedman`. I can even write a unit test to check that `__iter__` is idempotent.

Reply:
This approach has been agreed upon.

Setup is performed in the `__iter__` method so a new  `for` loop will generate the same data (is the random number generator is seeded).

For the corner case where we need to continue from where we stop the loop we can create a generator as follows:

```python
stream = iter(synt_dataset)

# Generate and print 3 samples
for i, (x,y) in enumerate(stream):
    print(x,y)
    if i == 3:
        break

# Calling again the "stream" generator will continue from it left
for i, (x,y) in enumerate(stream):
    print(x,y)
    if i == 3:
        break
```